sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java

```diff
@@ -488,22 +499,38 @@ private Schema buildNullSchema(Schema schema) {
     return kvs.setCoder(KvCoder.of(coder.getKeyCoder(), valueCoder));
   }
 
-  private List<Pair<Integer, Integer>> extractJoinColumns(int leftRowColumnCount) {
+  private static Field getFieldBasedOnRexNode(
+      Schema schema, RexNode rexNode, int leftRowColumnCount) {
+    if (rexNode instanceof RexInputRef) {
+      return schema.getField(((RexInputRef) rexNode).getIndex() - leftRowColumnCount);
+    } else if (rexNode instanceof RexFieldAccess) {
+      // need to extract field of Struct/Row.
+      RexFieldAccess fieldAccess = (RexFieldAccess) rexNode;
+      RexInputRef inputRef = (RexInputRef) fieldAccess.getReferenceExpr();
```

Comment:
Is the nested field access possible here? If it is, would the `fieldAccess.getReferenceExpr()` return `RexFieldAccess`?

Reply:
You are absolutely right. `row.row.row.field` should have `ref.fc.fc`. Improved to support longer field access chain.